### DISCUSSION TITLE
Why is there a huge `DOMElements` side-effect in the NPM bundle?

### DISCUSSION THREAD

**Discussion by danieltroger** [solidjs/solid#1036] [Q&A] [answered]
In my output code bundle after building my solid project there's a huge, unused, `new Set(["html","base","head","link","meta","style","title","body","address","article","aside","footer","header","main","nav","section","body","blockquote","dd","div","dl","dt","figcaption","figure","hr","li","ol","p","pre","ul","a","abbr","b","bdi","bdo","br","cite","code","data","dfn","em","i","kbd","mark","q","rp","rt","ruby","s","samp","small","span","strong","sub","sup","time","u","var","wbr","area","audio","img","map","track","video","embed","iframe","object","param","picture","portal","source","svg","math","canvas","noscript","script","del","ins","caption","col","colgroup","table","tbody","td","tfoot","th","thead","tr","button","datalist","fieldset","form","input","label","legend","meter","optgroup","option","output","progress","select","textarea","details","dialog","menu","summary","details","slot","template","acronym","applet","basefont","bgsound","big","blink","center","content","dir","font","frame","frameset","hgroup","image","keygen","marquee","menuitem","nobr","noembed","noframes","plaintext","rb","rtc","shadow","spacer","strike","tt","xmp","a","abbr","acronym","address","applet","area","article","aside","audio","b","base","basefont","bdi","bdo","bgsound","big","blink","blockquote","body","br","button","canvas","caption","center","cite","code","col","colgroup","content","data","datalist","dd","del","details","dfn","dialog","dir","div","dl","dt","em","embed","fieldset","figcaption","figure","font","footer","form","frame","frameset","head","header","hgroup","hr","html","i","iframe","image","img","input","ins","kbd","keygen","label","legend","li","link","main","map","mark","marquee","menu","menuitem","meta","meter","nav","nobr","noembed","noframes","noscript","object","ol","optgroup","option","output","p","param","picture","plaintext","portal","pre","progress","q","rb","rp","rt","rtc","ruby","s","samp","script","section","select","shadow","slot","small","source","spacer","span","strike","strong","style","sub","summary","sup","table","tbody","td","template","textarea","tfoot","th","thead","time","title","tr","track","tt","u","ul","var","video","wbr","xmp","input"]);`.

This comes from solid-js/web/dist/web.js where it's called `DOMElements` and exported. I can't find it anywhere in the code of this repo? What is it?

Can you change `const DOMElements = new Set(`

to `const DOMElements = /*@__PURE__*/ new Set(` so that terser (my minifier) can safely remove it?

**Comment by ryansolid** [comment, marked as answer]
Yeah this was added for hybrid universal rendering. It also may be used by some of the non-JSX templating. That being said I would have expected most bundlers to drop this if it is never imported but adding pure seems pretty easy to do.

**Reply by danieltroger** [reply]
Ok, great to know!
Yeah, terser doesn't drop it because you _could_ have defined `window.Set` to something else and then dropping it would change the behavior of the code. Would appreciate a pure comment :)

BTW (out of pure lazyness asking here) have you actually benchmarked the default JSX runtime that does `.innerHTML` on the `<template>` elements vs one that does `document.createElement`?

**Reply by ryansolid** [reply]
`document.createElement` is faster than `innerHTML` but `element.cloneNode` is faster than all.  I started with `document.createElement` and then went to using templates. Yes there is an upfront cost of creating the templates although it is not as much as you'd expect since those nodes aren't attached to the DOM. But cloning after the fact yields great benefits when you need them, like when you are rendering a number of elements on navigation. The larger the template the more impactful the performance improvement.

I briefly debated using `document.createElement` to create the nodes to clone but the code size explosion wouldn't be worth it I think.

**Reply by danieltroger** [reply]
Thank you for the response! I've been thinking about this before but never dared to ask haha.
Yeah tbh `document.createElement` to create them and then `.cloneNode` sounds like the perfect solution. Why would it be so much bigger? Isn't there already a `document.createElement` JSX runtime for solid-js lying around on some documentation page as example? Does using that disable usage of `Element.cloneNode()`?

**Reply by ryansolid** [reply]
You mean like the HyperScript one. Yeah but that one has other disadvantages too like needing to wrap expressions manually and it can't do other things we do like group computations. There are a number of optimizations and ergonomic improvements our custom compiler enables.

**Reply by danieltroger** [reply]
No, I meant the runtime mentioned here: https://github.com/solidjs/solid/releases/tag/v1.2.0
 
Ok idk about wrapping expressions and what group computations are. Also haven't looked into how solid's custom JSX compiler works. I guess I'll learn more as I use solid :)

**Reply by danieltroger** [reply]
@ryansolid did you add the pure comment or should I open an issue as reminder? 👀

**Reply by ryansolid** [reply]
Should have been fixed in the 1.4.4 release last week.

**Reply by danieltroger** [reply]
Amazing, tysm!
